aidial_adapter_vertexai/universal_api/storage.py

```diff
@@ -0,0 +1,142 @@
+        if self.bucket is None:
+            async with session.get(
+                f"{self.dial_url}/v1/bucket",
+                headers=self.auth.headers,
+            ) as response:
+                response.raise_for_status()
+                body = await response.json()
+                self.bucket = body["bucket"]
+
+        return self.bucket
+
+    @staticmethod
+    def _to_form_data(
+        filename: str, content_type: str, content: bytes
+    ) -> aiohttp.FormData:
+        data = aiohttp.FormData()
+        data.add_field(
+            "file",
+            io.BytesIO(content),
+            filename=filename,
+            content_type=content_type,
+        )
+        return data
+
+    async def upload(
+        self, filename: str, content_type: str, content: bytes
+    ) -> FileMetadata:
+        async with aiohttp.ClientSession() as session:
+            bucket = await self._get_bucket(session)
+            data = FileStorage._to_form_data(filename, content_type, content)
+            ext = _get_extension(content_type) or ""
+            url = f"{self.dial_url}/v1/files/{bucket}/{self.upload_dir}/{filename}{ext}"
+
+            async with session.put(
+                url=url,
+                data=data,
+                headers=self.auth.headers,
+            ) as response:
+                response.raise_for_status()
+                meta = await response.json()
+                logger.debug(f"Uploaded file: url={url}, metadata={meta}")
+                return meta
+
+    async def upload_file_as_base64(
+        self, data: str, content_type: str
+    ) -> FileMetadata:
+        filename = _compute_hash_digest(data)
+        content: bytes = base64.b64decode(data)
+        return await self.upload(filename, content_type, content)
+
+    def attachment_link_to_url(self, link: str) -> str:
+        base_url = f"{self.dial_url}/v1/files/"
+        return urljoin(base_url, link)
+
+    async def download_file_as_base64(self, url: str) -> str:
+        headers: Mapping[str, str] = {}
+        if url.startswith(self.dial_url):
+            headers = self.auth.headers
+
+        return await download_file_as_base64(url, headers)
+
+
+async def _download_file(url: str, headers: Mapping[str, str] = {}) -> bytes:
+    async with aiohttp.ClientSession() as session:
+        async with session.get(url, headers=headers) as response:
+            response.raise_for_status()
+            return await response.read()
+
+
+async def download_file_as_base64(
+    url: str, headers: Mapping[str, str] = {}
+) -> str:
+    bytes = await _download_file(url, headers)
+    return base64.b64encode(bytes).decode("ascii")
+
+
+def _compute_hash_digest(file_content: str) -> str:
+    return hashlib.sha256(file_content.encode()).hexdigest()
+
+
+def _get_extension(content_type: str) -> Optional[str]:
+    if content_type.startswith("image/"):
+        return "." + content_type[len("image/") :]
```

Comment:
Maybe this one would be safer to use?
https://docs.python.org/3/library/mimetypes.html#mimetypes.guess_extension

Reply:
Indeed. Fixed.